src/Winton.Extensions.Configuration.Consul/ConsulConfigurationProvider.cs

```diff
@@ -30,13 +33,18 @@ internal sealed class ConsulConfigurationProvider : ConfigurationProvider
 
             if (source.ReloadOnChange)
             {
-                ChangeToken.OnChange(
-                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _source.CancellationToken),
+                _cancellationTokenSource = new CancellationTokenSource();
+                _changeTokenRegistration = ChangeToken.OnChange(
+                    () => _consulConfigClient.Watch(_source.Key, _source.OnWatchException, _cancellationTokenSource.Token),
                     async () =>
                     {
                         await DoLoad(true).ConfigureAwait(false);
                         OnReload();
                     });
+                if (_source.CancellationToken.CanBeCanceled)
+                {
+                    _source.CancellationToken.Register(_cancellationTokenSource.Cancel);
+                }
```

Comment:
This is the bit I am not happy with yet. Either I'd like to refactor it to stop using a ChangeToken and just have a background task that's fired off [like the KeyVault provider does](https://github.com/aspnet/Extensions/blob/master/src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs#L130), or clean things up so we rely solely on the cancellation of the chagne token registration [like the file provider does](https://github.com/aspnet/Extensions/blob/master/src/Configuration/Config.FileExtensions/src/FileConfigurationProvider.cs#L35).